token/program-2022/src/extension/group_member_pointer/instruction.rs

```diff
@@ -0,0 +1,147 @@
+#[cfg(feature = "serde-traits")]
+use serde::{Deserialize, Serialize};
+use {
+    crate::{
+        check_program_account,
+        instruction::{encode_instruction, TokenInstruction},
+    },
+    bytemuck::{Pod, Zeroable},
+    num_enum::{IntoPrimitive, TryFromPrimitive},
+    solana_program::{
+        instruction::{AccountMeta, Instruction},
+        program_error::ProgramError,
+        pubkey::Pubkey,
+    },
+    spl_pod::optional_keys::OptionalNonZeroPubkey,
+    std::convert::TryInto,
+};
+
+/// Group member pointer extension instructions
+#[cfg_attr(feature = "serde-traits", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "serde-traits", serde(rename_all = "camelCase"))]
+#[derive(Clone, Copy, Debug, PartialEq, IntoPrimitive, TryFromPrimitive)]
+#[repr(u8)]
+pub enum GroupMemberPointerInstruction {
+    /// Initialize a new mint with a group member pointer
+    ///
+    /// Fails if the mint has already been initialized, so must be called before
+    /// `InitializeMint`.
+    ///
+    /// The mint must have exactly enough space allocated for the base mint (82
+    /// bytes), plus 83 bytes of padding, 1 byte reserved for the account type,
+    /// then space required for this extension, plus any others.
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   0. `[writable]` The mint to initialize.
+    ///   1. `[]`         The group mint.
+    ///   2. `[signer]`   The group's update authority.
```

Comment:
Why are these required? This should work just like the group pointer, right?

Reply:
I don't agree with this -- for the other pointer extensions, the pointer is always allowed to point anywhere, and it's up to the other extension to do the proper check, or for clients to make sure that the accounts point to each other.

Also, by requiring the authority check, this extension becomes incompatible with a token group that is owned by another program, which makes the pointer part unnecessary.

Reply:
I think I see what you're saying. The responsibility of validating a pointer is up to the program adding the extension being pointed to. In other words:
- `MetadataPointer`: Whoever creates `TokenMetadata` (Token2022 extension or elsewhere)
- `GroupPointer`: Whoever creates `TokenGroup` (Token2022 extension or elsewhere)
- `GroupMemberPointer`: Whoever creates `TokenGroupMember` (Token2022 extension or elsewhere)

Those programs are then responsible for validating one pointer over all the others. I think I got confused when it came to two related extensions. But I think we're on the same page now.

Reply:
I'd argue that the program being pointed to doesn't need to care either. It comes down to an agreement between two programs, which manifests as two accounts pointing at each other, ie a mint pointing at metadata, and metadata pointing at the mint. That's the *only* valid case, and everything else should be deemed invalid.

As metadata and collection programs proliferate, those programs can start doing whatever they want, but it's only "official" if the two point at each other.

You could make the argument that a metadata / collection program should validate the pointer before allowing the creation, but that could be limiting for some programs.